Objects/obmalloc.c

```diff
@@ -1477,20 +1477,18 @@ _PyMem_DebugRawRealloc(void *ctx, void *p, size_t nbytes)
         /* overflow:  can't represent total as a Py_ssize_t */
         return NULL;
 
+    if (nbytes < original_nbytes) {
+        /* shrinking:  mark old extra memory dead */
+        memset(q + nbytes, DEADBYTE, original_nbytes - nbytes + 2*SST);
```

Comment:
You cannot do that directly. If realloc() fails, you must restore original bytes.

Would it possible to allocate a new temporary memory block to save these erased bytes?

Reply:
It can require allocating large block.

This is why I added Py_FatalError(). If realloc() fails when shrinking memory block, something is wrong with memory allocator.

Reply:
> It can require allocating large block.

I'm not sure that it's a big issue. You should be prepared for memory allocation failures when you use debug hooks since they add a large overhead on memory footprint.

If you care, you might reimplement realloc as new=alloc()+memcpy(new,old)+free(old). "free(old)" will invalidate old bytes (*before* freeing memory) for you if I recall correctly.

Reply:
> If realloc() fails when shrinking memory block, something is wrong with memory allocator.

It is C-Standard conforming for `realloc()` to fail when shrinking the size. There are legitimate reasons to reshuffle blocks when shrinking (compact the heap). If that fails, `realloc()` *has* to signal to the caller that the requested size cannot be matched.

Reply:
> It is C-Standard conforming for `realloc()` to fail when shrinking the size.

I have not found an explicit specification of this behavior in the C Standard, but your arguments make sense to me.

Reply:
> I'm not sure that it's a big issue. You should be prepared for memory allocation failures when you use debug hooks since they add a large overhead on memory footprint.

This complicates the implementation. Could you create a new PR for your suggestion?

> If you care, you might reimplement realloc as new=alloc()+memcpy(new,old)+free(old). "free(old)" will invalidate old bytes (*before* freeing memory) for you if I recall correctly.

I don't like this. Always returning a new pointer can hide a common bug with misusing realloc API. Debug allocator should help to find bugs, not hide them.

Reply:
> I have not found an explicit specification of this behavior in the C Standard [...]

7.20.3.4 The realloc function

"The realloc function deallocates the old object pointed to by ptr and returns a
pointer to a new object that has the size specified by size."

"The realloc function returns a pointer to the new object (which may have the same
value as a pointer to the old object), or a null pointer if the new object could not be
allocated."

Reply:
Thank you @skrah.

Reply:
> This complicates the implementation. Could you create a new PR for your suggestion?

Done: PR #4119.